tsdb/head.go

```diff
@@ -1451,17 +1456,17 @@ type memSeries struct {
 
 	ref           uint64
 	lset          labels.Labels
-	mmappedChunks []*mmappedChunk
-	mmMaxTime     int64 // Max time of any mmapped chunk, only used during WAL replay.
-	headChunk     *memChunk
+	mmappedChunks []*mmappedChunk // Immutable chunks on disk that have not yet gone into a block, in order of ascending time stamps.
+	mmMaxTime     int64           // Max time of any mmapped chunk, only used during WAL replay.
+	headChunk     *memChunk       // Most recent chunk in memory that's still being built.
 	chunkRange    int64
-	firstChunkID  int
+	firstChunkID  int // ChunkID corresponding to mmappedChunks[0].ref
 
-	nextAt        int64 // Timestamp at which to cut the next chunk.
-	sampleBuf     [4]sample
-	pendingCommit bool // Whether there are samples waiting to be committed to this series.
+	nextAt        int64     // Timestamp at which to cut the next chunk.
+	sampleBuf     [4]sample // We keep the last 4 samples here (in addition to appending them) so we don't need coordination between appender and querier. Probably because even the most compact encoding of a sample takes 2 bits, so the last byte is not contended.
```

Comment:
Right about the sentence starting `Probably...` (but would not keep it in the code)
```suggestion
	// We keep the last 4 samples here (in addition to appending them to the chunk) so we don't
	// need coordination between appender and querier for the chunk bytes that are still being written to.
	sampleBuf     [4]sample
```

Reply:
> Right about the sentence starting Probably... (but would not keep it in the code)

I cleaned it up, but i think it's super useful to keep this in the code (without the "probably"), the question of "why do we keep exactly 4 samples in a separate buffer" was one of the great mysteries of TSDB to me, and the insight that we need to cover one byte and that 4 samples is what it takes, even in the most compact encoding case, is IMHO a big revelation that should be documented.

Reply:
The latest update makes sense, thanks